### PR TITLE
feat(generator): A basic yeoman provisioner generator

### PR DESCRIPTION
**Pull request by conneryn**
## Description

This is the start of a basic Yeoman Generator to create a new Provisioner from scratch.   There are a few improvements to be added to the code base, but happy to take code review suggestions now.

Items to consider and prioritize:
1. Refactor with Sub-Generators.  I'd like to separate different sections of code into separate sub-generators  (ex: have a `ui`, `k8s`, and `src-typescript` as separate sub-generators?).  However, I haven't decided what the right separation lines are.
2. Ensure the generated code is following our preferred conventions. Are there any specific coding guidelines we want to encourage developers to follow that we should include in the template?
3. Provisioner config to allow some changes after generating code.  We could factor out some of the yeoman variables that get injected all over the place (ex: applicationId) into an actual variable that on the generated code side (ex: create a config.ts that contains these?).  I'm not entirely convinced if this would make things easier or more confusing for external developers.
4. Sample code and comments.  Would it be helpful to have more "dummy" example code in the final generated code?
5. Other implementations.  At the very least, I'd like to make sure uninstall/remove actions are implemented as well.  But it may be worthwhile to add update, help, validate, and maybe even an exposed API.
6. Provisioner Tests. 
7. Generator Tests.

## Type of change

Please delete options that are not relevant.

- [ ] New feature (non-breaking change which adds functionality)
- [ ] This change requires a documentation update

## How Has This Been Tested?

Currently, this has only be tested locally and only using the CLI to provision.  Further testing is required.
NOTE: there is a bug on provisioning that needs to be resolved.

## Checklist:

- [ ] My code follows the style guidelines of this project
- [ ] I have performed a self-review of my own code
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [ ] My changes generate no new warnings
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] New and existing unit tests pass locally with my changes
- [ ] Any dependent changes have been merged and published in downstream modules
- [ ] I have checked my code and corrected any misspellings

### Node-RED
[asciicast](https://asciinema.org/a/lYDasi1nSZChrIktlNlMOIjvt)

### NextCloud
[asciicast](https://asciinema.org/a/D9PgjzrPzo2z3sXN4E7YJfxeY)

### CODE REVIEW

File: packages/generator-codezero-provisioner/LICENSE [hunk 1]
```diff
@@ -0,0 +1,13 @@
+Copyright 2020 Connery Noble <connery@codezero.io>
```
Comment (nsainaney) [inline packages/generator-codezero-provisioner/LICENSE]:
```suggestion
Copyright 2020 CodeZero Technologies Inc. Connery Noble <connery@codezero.io>
```

File: packages/generator-codezero-provisioner/README.md [hunk 2]
```diff
@@ -0,0 +1,36 @@
...
+## Getting To Know Yeoman
+
+ * Yeoman has a heart of gold.
+ * Yeoman is a person with feelings and opinions, but is very easy to work with.
+ * Yeoman can be too opinionated at times but is easily convinced not to be.
+ * Feel free to [learn more about Yeoman](http://yeoman.io/).
+
+## License
+
+Apache-2.0 © [Connery Noble]()
```
Comment (nsainaney) [inline packages/generator-codezero-provisioner/README.md]:
```suggestion
Apache-2.0 © [CodeZero Technologies Inc.](https://codezero.io)
by Connery Noble
```

File: packages/generator-codezero-provisioner/generators/app/index.js [hunk 3]
```diff
@@ -0,0 +1,178 @@
...
+const chalk = require("chalk");
+const yosay = require("yosay");
+const _ = require("lodash");
+
+module.exports = class extends Generator {
+  async prompting() {
+    // Have Yeoman greet the user.
+    this.log(
+      yosay(
+        `Welcome to the stupendous ${chalk.red(
```
Comment (nsainaney) [inline packages/generator-codezero-provisioner/generators/app/index.js]:
```suggestion
        `Welcome to the kubetastic ${chalk.red(
```

File: packages/generator-codezero-provisioner/generators/app/templates/README.md [hunk 4]
```diff
@@ -0,0 +1,13 @@
...
+This project is a provisioner to install and manage <%= applicationName %> in CodeZero.
+
+## Running This Locally
+
+To run this provisioner, you will need a Kubernetes cluster up and running, with CodeZero installed.  Head over to https://hub.codezero.io/ to get started.
+
+```
+export KUBECONFIG=/path/to/your/cluster-config.yaml
+npm install
+npx czctl provision app.yaml --package ./
```
Comment (nsainaney) [inline packages/generator-codezero-provisioner/generators/app/templates/README.md]: We tried `--package` the other day and it did not work for us. Also, it should not be in the provisioned manager where this was implemented but that's a minor todo

File: packages/generator-codezero-provisioner/generators/app/templates/app.yaml [hunk 5]
```diff
@@ -0,0 +1,23 @@
...
+  annotations:
+    system.codezero.io/display: "<%= applicationName %>"
+spec:
+  package: "<%= provisionerPackageName %>"
+  provisioner:
+    tag: "<%= tag %>"
+  routes:
+  - type: <%= serviceType %>
+    targetService: <%= applicationId %>
+<% if (serviceType == 'tcp') { -%>
```
Comment (nsainaney) [inline packages/generator-codezero-provisioner/generators/app/templates/app.yaml]: Most services will be http

File: packages/generator-codezero-provisioner/generators/app/index.js [hunk 6]
```diff
@@ -0,0 +1,178 @@
...
+        default: true
+      },
+      {
+        when: ({ pvcEnabled }) => pvcEnabled,
+        type: "input",
+        name: "pvcMountPath",
+        message: "Where should the persistence volume be mounted?",
+        default: "/var/data"
+      }
+    ]);
```
Comment (nsainaney) [inline packages/generator-codezero-provisioner/generators/app/index.js]: This stuff get's very provisioner specific. I'd almost leave the k8s folder blank

File: packages/generator-codezero-provisioner/generators/app/templates/src/index.ts [hunk 7]
```diff
@@ -0,0 +1,15 @@
+import { mix } from "mixwith"
+import { ProvisionerBase } from "@provisioner/common"
+
+import {
+    createApplyMixin,
+    createInquireMixin,
+    createValidateMixin,
+    removeApplyMixin,
```
Comment (nsainaney) [inline packages/generator-codezero-provisioner/generators/app/templates/src/index.ts:8]: These are optional - should prompt to see if the use needs them

File: packages/generator-codezero-provisioner/generators/app/templates/src/mixins/createApply.ts [hunk 8]
```diff
@@ -0,0 +1,65 @@
+import { baseProvisionerType } from "../index"
+
+export const createApplyMixin = (base: baseProvisionerType) => class extends base {
+
+    async createApply() {
+        await this.ensureServiceNamespacesExist()
+        await this.ensureApplicationIsInstalled()
+        await this.ensureApplicationIsRunning()
```
Comment (nsainaney) [inline packages/generator-codezero-provisioner/generators/app/templates/src/mixins/createApply.ts:8]: We should move ensure application is running into a helper function in the base class

File: packages/generator-codezero-provisioner/generators/app/templates/src/ui/install.ts [hunk 9]
```diff
@@ -0,0 +1,47 @@
...
+        return html`
+<% if (pvcEnabled) { -%>
+            <c6o-form-layout>
+                <c6o-combo-box @selected-item-changed=${this.storageSelected} label="<%= applicationName %> Storage" value=${this.serviceSpec.storage} required allow-custom-value .items=${this.values}></c6o-combo-box>
+            </c6o-form-layout>
+<% } else { -%>
+            <c6o-form-layout>
+                <c6o-text-field @input=${this.examplePropertyChanged} label="Example Property" value=${this.serviceSpec.exampleProperty || ''} autoselect></c6o-text-field>
+            </c6o-form-layout>
+<% } -%>
```
Comment (nsainaney) [inline packages/generator-codezero-provisioner/generators/app/templates/src/ui/install.ts]: I'd prompt if they want ui generated. If so, I'd leave the HTML blank or `insert controls here`

File: packages/generator-codezero-provisioner/generators/app/templates/src/index.ts (hunk 7 above)
Comment (conneryn) [reply packages/generator-codezero-provisioner/generators/app/templates/src/index.ts:8]: Sure, though I'd like to limit the number of prompts if possible (there are already a lot of questions).  Maybe we could have it as a commandline arg instead?  For example: `--excludeRemoveMixin`. (or --include)

Is there any harm in generating the scaffolding for these, and adding a comment in the method to explain it's entirely optional?

Comment (conneryn) [reply packages/generator-codezero-provisioner/generators/app/templates/src/index.ts:8]: Maybe we could have a single question that asks the user if they want stub code generated or not.  And then that would apply to ALL extra code, like removeApply, createValidate, ui, etc?   Or I guess could break it down even more, like:
(1) Would you like stub code generated for this project?
      - All
      - None
      - Custom

And if they select custom they would go through a bunch more questions about each specific code stub.

File: packages/generator-codezero-provisioner/generators/app/templates/src/mixins/createApply.ts (hunk 8 above)
Comment (conneryn) [reply packages/generator-codezero-provisioner/generators/app/templates/src/mixins/createApply.ts:8]: Yeah, that would be create!  Could we do the same for ensureNamespaceExists()?

File: packages/generator-codezero-provisioner/generators/app/templates/src/ui/install.ts (hunk 9 above)
Comment (conneryn) [reply packages/generator-codezero-provisioner/generators/app/templates/src/ui/install.ts]: Same topic as above about stub code.  Refer to discussion there.

File: packages/generator-codezero-provisioner/generators/app/index.js (hunk 6 above)
Comment (conneryn) [reply packages/generator-codezero-provisioner/generators/app/index.js]: I see what you mean, but I think we lose most of the benefit of scaffolding if we don't generate any of the k8s.  So much of our naming convention of things relies on variable coordination between the Application Spec, createApplyMixin, and the k8s Yamls.

If we don't create any k8s:
- then createApply upsertFiles won't exist anymore.
- Application specs like `routes, and `marina` won't be able to be generated.
- We lost the ability to encourage some recommended best practices in what the yamls look like (ex: they should inject {{namespace}})

I also think there is benefit to being able to create the ability to generate a zerocode provisioner out of the box (if a docker image is simple enough).  For example, this currently works with NodeRED, Registry, NextCloud.

File: packages/generator-codezero-provisioner/generators/app/templates/README.md (hunk 4 above)
Comment (conneryn) [reply packages/generator-codezero-provisioner/generators/app/templates/README.md]: I'd be curious what part of using `--package` isn't working for you?  Currently this is the only way I have that works without needing to build czctl from monorepo and muck around with PATHs.

Comment (conneryn) [reply packages/generator-codezero-provisioner/generators/app/templates/README.md]: This is also somewhat of a mute point, as I am hoping we can change all this very soon, once the new Package Manifest (Provisioner Manifest? ... Something Manifest), is more finalized.

**Comment by conneryn** [comment]
There is still lots of room for improvement on the template structure and questions that will should be asking.  Much of which will be dictated by the evolution of the AppEngine + Studio features and the provisioner package.  

But I think it would still be worth getting feedback on the other structures so far (ex: direction of unit testing, tslinting, and general inquire/generation structure, etc), and getting this merged in the mean time.

**Comment by conneryn** [comment]
Closing this PR until I've updated to be compatible with base provisioner changes, code style guide, and get closer to AppEngine parity.

**Pull request closed without merging**